README.md
# MacGyver

Duct Tape and a Swiss Army Knife. Angular helpers for all your friends!

[![Build Status](http://img.shields.io/travis/angular-macgyver/MacGyver/master.svg?style=flat-square)](https://travis-ci.org/angular-macgyver/MacGyver)
[![Coverage Status](https://img.shields.io/coveralls/angular-macgyver/MacGyver/master.svg?style=flat-square)](https://coveralls.io/r/angular-macgyver/MacGyver?branch=master)
[![License](http://img.shields.io/badge/license-MIT-green.svg?style=flat-square)](https://github.com/angular-macgyver/MacGyver/blob/master/LICENSE)
[![Latest Release](http://img.shields.io/github/release/angular-macgyver/MacGyver.svg?style=flat-square)](https://github.com/angular-macgyver/MacGyver/releases/latest)

[![Selenium Test Status](https://saucelabs.com/browser-matrix/macgyver-ci.svg)](https://saucelabs.com/u/macgyver-ci)

## Components ##

### Directives ###
 - Affix
 - Autocomplete
 - Canvas Spinner
 - Events
 - Menu
 - Modal
 - Placeholder
 - Popover
 - Scroll Spy
 - Spinner
 - Tag Autocomplete
 - Time Input
 - Tooltip

### Filters ###
 - Boolean
 - Pluralize
 - Timestamp

## 3rd party libraries dependencies ##
 - AngularJS (1.2.x+)

## Using MacGyver ##
You can install via [Bower](http://www.bower.io) or download from [Github](https://github.com/angular-macgyver/MacGyver/archive/master.zip)

To install via bower, make sure you have bower installed and then run:

```
bower install angular-macgyver
```

To install via NPM, you can run:

```
npm install angular-macgyver
```

To install via Yarn, you can run:
```
yarn add angular-macgyver
```

Once you have MacGyver in your project, just include "Mac" as a dependency in your Angular application and you’re good to go.

```javascript
angular.module("myModule", ["Mac"]);
```


CHANGELOG.md
# v1.0.0 (2016/10/8)
## Bug Fixes
- **macPopover:** Fix mac-popover-refresh-on not registering properly (ec9e4ff4, #38)

# 1.0.0-rc.1 (2016/5/22)
## Bug Fixes
- **macTagAutocomplete:**
  - Fix macTagAutocompleteOnSuccess not using '&'
  (50edb3a7)
  - Fix not creating event directives
  (f13e6bb0)

## Optimizations
- **macTagAutocomplete:**
  - Switch to use macTagAutocompleteController (9e83d0e1)
  - Add mac-tag-autocomplete-on-success attribute for callback on a successful xhr request. This will allow users to modify the data pass back to mac-autocomplete before rendering (9e83d0e1)

## Breaking Changes
- **macTagAutocomplete:**
  - Directly populate item from source without filtering (9e83d0e1)


# 1.0.0-rc.0 (2016/2/10)
## Bug Fixes
- **macAffix:** Fix referencing incorrect variable due to JS conversion (7d95b2a7)
- **macModal:** Fix not closing modal with escape (ac70db5c)
- **macPopover:**
  - Fix always switching to aligning right (34d94c5e)
  - Fix not extending proper popover default values (bf03f6b3)
- **macTagAutocomplete:** Fix bugs introduced in refactor (c55011d1)
- **macTooltip:** Fix appending tooltip inside does not calculate offset correctly (19cbb097, #31)
- **main:** Update height and width calculation to prevent NaN (d77e40db)

## Breaking Changes
- **macAutocomplete:**
  - macAutocomplete is now isolated scope
  - mac-autocomplete-url has been removed in favor of mac-autocomplete-source (438e8fe6)
- **macCSpinner:**  Canvas spinner has been removed. Use macSpinner instead. (f6d04ce3)
- **macFocusOnEvent:**  mac-focus-on-event directive has been removed
from MacGyver. The directive was never documented and should not be
used. (f0e12ecd)
- **macPopover:**
  - mac-popover-fixed option has been removed (7ef8a197)
  - macPopover service `resize` renamed to `reposition`
  - mac-popover-child-popover attr has been removed
  - mac-popover-exclude attr has been removed (2b3c91e6)
- **macTagAutocomplete:**
  - mac-tag-autocomplete-url has been removed (fa4bab82)
- **underscoreStringFilter:**  Underscore string filter has been completely removed (c9ec969a)

## Optimizations
- **booleanFilter:** Converted to Javascript (8a7d3149)
- **constants:** Move component default values to constants file (7222c105)
- **keys:** Converted keys to Javascript Switched from factory to constants (7467e789)
- **listFilter:** Converted to Javascript (89423980)
- **macAffix:**
  - Clean up affix directive Move logic to macAffixController Fix some bus causing display issues (48b774e8)
  - Add macAffixDefaults (4f94a667)
  - Converted to Javascript (6b106c94)
- **macAutocomplete:** Embed template into directive js (29463577)
- **macKeydown:** Converted to Javascript (c7f7b509)
- **macMenu:** Converted to Javascript (91cf2c95)
- **macModal:**
  - Clean up mac-modal trigger (451c2a0a)
  - Convert macModal to JS - Move macModal defaults to constants - Clean up template, render and display modal logic - Add unit tests for modal service (673e1b86)
- **macPauseTyping:** Converted to Javascript (73ce8384)
- **macPlaceholder:** Converted to Javascript (4ee11131)
- **macScrollSpy:**
  - Clean up observe logic on anchor and target (873240c7)
  - Converted to Javascript Minor performance improvements (5bbbab09)
- **macSpinner:** Converted to Javascript (6dcfa1c0)
- **macTime:**
  - Embed template into directive js (621d371d)
  - Convert to Javascript Move time utility function to a separate service for better testing Add tests for time utility function (bd96c7d4)
- **macTooltip:** Convert to Javascript (95287ace)
- **macUtil:**
  - Cleaned up JS code (18298981)
  - Converted to Javascript Added `validateTime` function urlRegex, emailRegex and timeRegex are switched to local variables Updated macTime using `validateTime` (04c0ba3d)
- **macWindowResize:** Minor performance tweak and code cleanup Converted to Javascript and rewrote part of the directive (bba5ebc7)
- **main:** Converted to Javascript Removed ngAnimate dependencies for Mac Updated karma config (86933886)
- **pluralizeFilter:** Converted to Javascript (c3b5cf78)
- **timestampFilter:** Converted to Javascript Cleaned up time calculation (c6d9adf3)


# v0.6.1 (2015/2/6)
## Bug Fixes
- **macAutocomplete:** Fixed dropdown placement issues
Related to #11
  ([725995cb](https://github.com/angular-macgyver/MacGyver/commit/725995cb54a171d21b41c1190ce246e0a1b5f067))


# v0.6.0 (2014/12/4)
## Features
- Support AngularJS v1.3+

## Bug Fixes
- **macAutocomplete:** Fixed $animate API changes due to AngularJS v1.3
([d6c69cd7](https://github.com/angular-macgyver/MacGyver/commit/d6c69cd7dedbf0ef3d8620b4b0be29339c8a1a9e))
- **macModal:** Fixed not able to close with escape key
([373d6647](https://github.com/angular-macgyver/MacGyver/commit/373d6647c00e6c02eed374618cbc0b3d8594393b))

## Optimizations
- **Mac:** Initial 1.3 changes
([58207495](https://github.com/angular-macgyver/MacGyver/commit/582074950b5f034b3a8c4839534f5888f9c8821c))

## Breaking Changes
- **macModal:**  modalView provider has been removed and should be switched to use modalProvider instead  Related to #4
([c5fb9815](https://github.com/angular-macgyver/MacGyver/commit/c5fb98153e9f4f423019de545decb20f2cf1f274))
- **macPopover:**  popoverView provider has been removed and should be switched to use popoverProvider instead  Related to #4
([f38cb92f](https://github.com/angular-macgyver/MacGyver/commit/f38cb92f132ccc81d1ef6501e02421cb412e984e))


# v0.5.1 (2014/12/2)
## Bug Fixes
- **macTooltip:** Fixed appending multiple tooltips
Multiple tooltips get appended to document body when user mouse over the tooltip trigger multiple in a short period of time
([8070c262](https://github.com/angular-macgyver/MacGyver/commit/8070c2622de20b2f5497344791868dafa7af0ab9))


# v0.5.0 (2014/11/13)
## Bug Fixes
- **core:** Fixed val not getting initialized correctly
  ([c6eba169](https://github.com/angular-macgyver/MacGyver/commit/c6eba16908a9c73b36b19cf6b909f44af65f41e4))

lib/angular-util.js

angular.module("Util", []);

angular.module("Util").factory("Util.awesome", function() {
  return "awesome";
});
var event, key, _fn, _fn1, _i, _j, _len, _len1, _ref, _ref1;

_ref = ["Blur", "Focus", "Keydown", "Keyup", "Mouseenter", "Mouseleave"];
_fn = function(event) {
  return angular.module("Util").directive("util" + event, [
    "$parse", function($parse) {
      return {
        restrict: "A",
        link: function(scope, element, attributes) {
          var expression;
          expression = $parse(attributes["util" + event]);
          return element.on(event.toLowerCase(), function(event) {
            scope.$apply(function() {
              return expression(scope, {
                $event: event
              });
            });
            return true;
          });
        }
      };
    }
  ]);
};
for (_i = 0, _len = _ref.length; _i < _len; _i++) {
  event = _ref[_i];
  _fn(event);
}

_ref1 = ["Enter", "Escape", "Space", "Left", "Up", "Right", "Down"];
_fn1 = function(key) {
  return angular.module("Util").directive("utilKeydown" + key, [
    "$parse", "Util.keys", function($parse, keys) {
      return {
        restrict: "A",
        link: function(scope, element, attributes) {
          var expression;
          expression = $parse(attributes["utilKeydown" + key]);
          return element.on("keydown", function(event) {
            if (event.which === keys["" + (key.toUpperCase())]) {
              event.preventDefault();
              return scope.$apply(function() {
                return expression(scope, {
                  $event: event
                });
              });
            }
          });
        }
      };
    }
  ]);
};
for (_j = 0, _len1 = _ref1.length; _j < _len1; _j++) {
  key = _ref1[_j];
  _fn1(key);
}

angular.module("Util").directive("utilModelBlur", [
  "$parse", function($parse) {
    return {
      restrict: "A",
      require: "ngModel",
      link: function(scope, element, attributes, controller) {
        return element.on("blur", function(event) {
          if (controller.$valid) {
            return scope.$apply($parse(attributes.utilModelBlur)(scope, {
              $event: event
            }));
          }
        });
      }
    };
  }
]);

angular.module("Util").directive("utilPauseTyping", [
  "$parse", function($parse) {
    return {
      restrict: "A",
      link: function(scope, element, attributes) {
        var delay, expression, keyupTimer;
        expression = $parse(attributes["utilPauseTyping"]);
        delay = scope.$eval(attributes["utilPauseTypingDelay"]) || 800;
        keyupTimer = "";
        return element.on("keyup", function(event) {
          clearTimeout(keyupTimer);
          return keyupTimer = setTimeout((function() {
            return scope.$apply(function() {
              return expression(scope, {
                $event: event
              });
            });
          }), delay);
        });
      }
    };
  }
]);

angular.module("Util").factory("Util.keys", function() {
  return {
    CANCEL: 3,
    HELP: 6,
    BACKSPACE: 8,
    TAB: 9,
    CLEAR: 12,
    ENTER: 13,
    RETURN: 13,
    SHIFT: 16,
    CONTROL: 17,
    ALT: 18,
    PAUSE: 19,
    CAPS_LOCK: 20,
    ESCAPE: 27,
    SPACE: 32,
    PAGE_UP: 33,
    PAGE_DOWN: 34,
    END: 35,
    HOME: 36,
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
    PRINT_SCREEN: 44,
    INSERT: 45,
    DELETE: 46,
    ZERO: 48,
    ONE: 49,
    TWO: 50,
    THREE: 51,
    FOUR: 52,
    FIVE: 53,
    SIX: 54,
    SEVEN: 55,
    EIGHT: 56,
    NINE: 57,
    SEMICOLON: $.browser.safari ? 186 : 59,
    EQUALS: 61,
    A: 65,
    B: 66,
    C: 67,
    D: 68,
    E: 69,
    F: 70,
    G: 71,
    H: 72,
    I: 73,
    J: 74,
    K: 75,
    L: 76,
    M: 77,
    N: 78,
    O: 79,
    P: 80,
    Q: 81,
    R: 82,
    S: 83,
    T: 84,
    U: 85,
    V: 86,
    W: 87,
    X: 88,
    Y: 89,
    Z: 90,
    CONTEXT_MENU: 93,
    NUMPAD0: 96,
    NUMPAD1: 97,
    NUMPAD2: 98,
    NUMPAD3: 99,
    NUMPAD4: 100,
    NUMPAD5: 101,
    NUMPAD6: 102,
    NUMPAD7: 103,
    NUMPAD8: 104,
    NUMPAD9: 105,
    MULTIPLY: 106,
    ADD: 107,
    SEPARATOR: 108,
    SUBTRACT: 109,
    DECIMAL: 110,
    DIVIDE: 111,
    F1: 112,
    F2: 113,
    F3: 114,
    F4: 115,
    F5: 116,
    F6: 117,
    F7: 118,
    F8: 119,
    F9: 120,
    F10: 121,
    F11: 122,
    F12: 123,
    F13: 124,
    F14: 125,
    F15: 126,
    F16: 127,
    F17: 128,
    F18: 129,
    F19: 130,
    F20: 131,
    F21: 132,
    F22: 133,
    F23: 134,
    F24: 135,
    NUM_LOCK: 144,
    SCROLL_LOCK: 145,
    COMMA: 188,
    PERIOD: 190,
    SLASH: 191,
    BACK_QUOTE: 192,
    OPEN_BRACKET: 219,
    BACK_SLASH: 220,
    CLOSE_BRACKET: 221,
    QUOTE: 222,
    META: 224
  };
});
